test/hotspot/gtest/runtime/test_os_linux.cpp

```diff
@@ -350,6 +350,70 @@ TEST_VM(os_linux, reserve_memory_special_concurrent) {
   }
 }
 
+class PretouchMemoryRunnable : public TestRunnable {
+  char* addr;
+  size_t byte;
+public:
+  PretouchMemoryRunnable(char* addr, size_t byte): addr(addr), byte(byte) {}
+
+  void runUnitTest() const {
+    os::pretouch_memory(addr, addr + byte, os::vm_page_size());
+  }
+};
+
+class UseMemoryRunnable : public TestRunnable {
+  char* addr;
+  size_t byte;
+public:
+  UseMemoryRunnable(char* addr, size_t byte): addr(addr), byte(byte) {}
+
+  void runUnitTest() const {
+    int* iptr = reinterpret_cast<int*>(addr);
+    for (int i = 0; i < 1000 && (size_t)i < (byte / (sizeof(int))); i++)
+      *iptr++ = i;
+  }
+};
+
+static void *pthread_test_runnable(void *runnable) {
+  TestRunnable* r = reinterpret_cast<TestRunnable*>(runnable);
+  r->runUnitTest();
+  return nullptr;
+}
+
+TEST_VM(os_linux, pretouch_thp_and_use_concurrent) {
+  // Explicitly enable thp to test cocurrent system calls.
+  bool useThp = UseTransparentHugePages;
+  UseTransparentHugePages = true;
+  char* heap = os::reserve_memory(1 * G, false, mtInternal);
+  EXPECT_NE(heap, (char*)NULL);
+  EXPECT_TRUE(os::commit_memory(heap, 1 * G, false));
+
+  {
+    UseMemoryRunnable use(heap, 1 * G);
+    PretouchMemoryRunnable pretouch(heap, 1 * G);
+    pthread_t threads[8];
+    pthread_attr_t attributes[8];
+    for (int i = 0; i < 8; i++) {
+      pthread_attr_init(attributes + i);
+      TestRunnable *runnable = &pretouch;
+      if (i == 0) runnable = &use;
+      pthread_create(threads + i, attributes + i, pthread_test_runnable, runnable);
+      pthread_attr_destroy(attributes + i);
+    }
+    for (int i = 0; i < 8; i++) {
+      pthread_join(threads[i], 0);
+    }
+  }
+
```

Comment:
Is there a reason that you want to use pthreads directly instead of using `TestThreadGroup`?

Draft:

```c++
char* heap = os::reserve_memory(1 * G, false, mtInternal);
size_t size = 1*G;
auto pretouch = [&](Thread* c, int id)) {
  os::pretouch_memory(heap, heap + byte, os::vm_page_size());
}; 
auto use_memory = [&](Thread* c, int id) {
    int* iptr = reinterpret_cast<int*>(heap);
    for (int i = 0; i < 1000 && (size_t)i < (byte / (sizeof(int))); i++)
      *iptr++ = i;
};
TestThreadGroup<decltype(use_memory)> users_t(use_memory, 8);
TestThreadGroup<decltype(pretouch)>  pretouch_t(pretouch, 1);
users_t.doit();
pretouch_t.doit();
users_t.join();
pretouch_t.join();
```